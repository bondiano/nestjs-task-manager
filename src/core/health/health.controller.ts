import os from 'node:os';

import { Controller, Get } from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import {
  HealthCheck,
  HealthCheckService,
  MemoryHealthIndicator,
  TypeOrmHealthIndicator,
} from '@nestjs/terminus';

@Controller('health')
@ApiTags('Health Check')
export class HealthController {
  constructor(
    private health: HealthCheckService,
    private database: TypeOrmHealthIndicator,
    private memory: MemoryHealthIndicator
  ) {}

  @Get()
  @HealthCheck()
  check() {
    const totalMemory = os.totalmem();

    return this.health.check([
      () => this.database.pingCheck('database'),
      () =>
        this.memory.checkRSS(
          'memory_rss',
          totalMemory * 0.9 /* 90% of total memory */
        ),
    ]);
  }
}
