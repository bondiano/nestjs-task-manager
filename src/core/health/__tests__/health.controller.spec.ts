import {
  HealthCheckService,
  HealthIndicatorFunction,
  MemoryHealthIndicator,
  TypeOrmHealthIndicator,
} from '@nestjs/terminus';
import { Test, TestingModule } from '@nestjs/testing';

import { HealthController } from '../health.controller';

const getStatus = (key: string) => ({ [key]: { status: 'up' } });

describe('HealthController', () => {
  let controller: HealthController;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [HealthController],
    })
      .useMocker((token) => {
        if (Object.is(token, HealthCheckService)) {
          return {
            check: jest
              .fn()
              .mockImplementation(
                (indicators: Array<HealthIndicatorFunction>) =>
                  Promise.all(indicators.map((indicator) => indicator()))
              ),
          };
        }

        if (Object.is(token, TypeOrmHealthIndicator)) {
          return {
            pingCheck: jest.fn().mockImplementation(getStatus),
          };
        }

        if (Object.is(token, TypeOrmHealthIndicator)) {
          return {
            pingCheck: jest.fn().mockImplementation(getStatus),
          };
        }

        if (Object.is(token, MemoryHealthIndicator)) {
          return {
            checkHeap: jest.fn().mockImplementation(getStatus),
            checkRSS: jest.fn().mockImplementation(getStatus),
          };
        }
      })
      .compile();

    controller = module.get<HealthController>(HealthController);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  it('should check health', async () => {
    await expect(controller.check()).resolves.toMatchSnapshot();
  });
});
