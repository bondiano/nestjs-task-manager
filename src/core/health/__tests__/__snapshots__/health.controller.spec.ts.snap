// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthController should check health 1`] = `
[
  {
    "database": {
      "status": "up",
    },
  },
  {
    "memory_rss": {
      "status": "up",
    },
  },
]
`;
